src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -364,45 +451,53 @@ public void OnResponseHeadersComplete(bool endStream)
                     {
                         return;
                     }
 
-                    if (_state != StreamState.ExpectingHeaders && _state != StreamState.ExpectingTrailingHeaders && _state != StreamState.ExpectingIgnoredHeaders)
+                    if (_responseProtocolState != ResponseProtocolState.ExpectingHeaders && _responseProtocolState != ResponseProtocolState.ExpectingTrailingHeaders && _responseProtocolState != ResponseProtocolState.ExpectingIgnoredHeaders)
                     {
                         throw new Http2ConnectionException(Http2ProtocolErrorCode.ProtocolError);
                     }
 
-                    if (_state == StreamState.ExpectingHeaders)
+                    if (_responseProtocolState == ResponseProtocolState.ExpectingHeaders)
                     {
-                        _state = endStream ? StreamState.Complete : StreamState.ExpectingData;
+                        _responseProtocolState = endStream ? ResponseProtocolState.Complete : ResponseProtocolState.ExpectingData;
                     }
-                    else if (_state == StreamState.ExpectingTrailingHeaders)
+                    else if (_responseProtocolState == ResponseProtocolState.ExpectingTrailingHeaders)
                     {
                         if (!endStream)
                         {
                             if (NetEventSource.IsEnabled) Trace("Trailing headers received without endStream");
                             throw new Http2ConnectionException(Http2ProtocolErrorCode.ProtocolError);
                         }
 
-                        _state = StreamState.Complete;
+                        _responseProtocolState = ResponseProtocolState.Complete;
                     }
-                    else if (_state == StreamState.ExpectingIgnoredHeaders)
+                    else if (_responseProtocolState == ResponseProtocolState.ExpectingIgnoredHeaders)
                     {
                         if (endStream)
                         {
                             // we should not get endStream while processing 1xx response.
                             throw new Http2ConnectionException(Http2ProtocolErrorCode.ProtocolError);
                         }
 
-                        _state = StreamState.ExpectingStatus;
+                        _responseProtocolState = ResponseProtocolState.ExpectingStatus;
                         // We should wait for final response before signaling to waiter.
                         return;
                     }
                     else
                     {
-                        _state = StreamState.ExpectingData;
+                        _responseProtocolState = ResponseProtocolState.ExpectingData;
+                    }
+
+                    if (endStream)
+                    {
+                        Debug.Assert(_responseCompletionState == StreamCompletionState.InProgress, $"Response already completed with state={_responseCompletionState}");
```

Comment:
Can we assert this safely? i.e. what if the server sent erroneous data... could it trigger this assert?

Reply:
When we process the first endStream, we should set _responseProtocolState (NOT _responseCompletionState) to Completed. Then any subsequent frames received should throw a protocol error when they check the state. So we shouldn't get here.

That said, some of the state logic around _responseProtocolState above and elsewhere is a little messy, so I can't say 100% we catch all of these. We have tests for this, but we could be missing something. I've tried not to mess with _responseProtocolState too much here because I didn't want to break anything here, but perhaps I should review all this and clean it up a bit.